### PR TITLE
TF-1362: Query mailbox spam report

### PR DESCRIPTION
**Pull request by nqhhdev**
TF-1362: Query mailbox with role spam report

### CODE REVIEW

File: lib/jmap/mail/mailbox/query/query_mailbox_method.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+      }
+    }
+    
+    writeNotNull('filter', filter?.toJson());
+    writeNotNull('limit', const UnsignedIntNullableConverter().toJson(limit));
+    return val;
+  }
+}
+
+mixin OptionalCollapseThreads {
```
Comment (dab246) [inline lib/jmap/mail/mailbox/query/query_mailbox_method.dart:56]: As jmap docs, we not need it. Please read https://jmap.io/spec-mail.html#mailboxquery

**Review by dab246 — COMMENTED** [review]
Screenshot 2023-01-10 at 13 03 04

You should implement all properties like doc to use it in future

File: lib/jmap/mail/mailbox/mailbox_filter_condition.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
+
+import 'package:jmap_dart_client/http/converter/utc_date_nullable_converter.dart';
+import 'package:jmap_dart_client/jmap/core/filter/filter_condition.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'mailbox_filter_condition.g.dart';
+
+@UTCDateNullableConverter()
```
Comment (dab246) [inline lib/jmap/mail/mailbox/mailbox_filter_condition.dart:8]: No need

File: test/jmap/mailbox/query/query_mailbox_test.dart [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+import 'package:jmap_dart_client/jmap/jmap_request.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/get/get_mailbox_method.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/get/get_mailbox_response.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox_filter_condition.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/mailbox_rights.dart';
+import 'package:jmap_dart_client/jmap/mail/mailbox/query/query_mailbox_method.dart';
+
+void main(){
+  group('Query mailbox', () {
```
Comment (dab246) [inline test/jmap/mailbox/query/query_mailbox_test.dart:21]: `'Query mailbox test'`

**Review by dab246 — COMMENTED** [review]
Screenshot 2023-01-10 at 14 08 44

Missing 2 optionals for `Mailbox/query`

File: test/jmap/mailbox/query/query_mailbox_test.dart [hunk 4]
```diff
@@ -0,0 +1,169 @@
+
+
```
Comment (dab246) [inline test/jmap/mailbox/query/query_mailbox_test.dart:2]: remove line

File: lib/jmap/mail/mailbox/mailbox_filter_condition.dart [hunk 5]
```diff
@@ -1,19 +1,41 @@
...
+@HasAnyRoleConverter()
 class MailboxFilterCondition extends FilterCondition {
 
   @JsonKey(includeIfNull: false)
-  final String? role;
+  final Role? role;
+  @JsonKey(includeIfNull: false)
+  final MailboxName? name;
+  @JsonKey(includeIfNull: false)
+  final HasAnyRole? hasAnyRole;
```
Comment (hoangdat) [inline lib/jmap/mail/mailbox/mailbox_filter_condition.dart:27]: why not use `bool?`?

File: lib/jmap/core/method/request/query_method.dart [hunk 6]
```diff
@@ -8,7 +8,8 @@ import 'package:json_annotation/json_annotation.dart';
 
 abstract class QueryMethod extends MethodRequiringAccountId
     with OptionalPosition, OptionalAnchorOffset, OptionalCalculateTotal,
-        OptionalFilter, OptionalSort, OptionalAnchor, OptionalLimit {
+        OptionalFilter, OptionalSort, OptionalAnchor, OptionalLimit, FilterAsTree,
+        SortAsTree {
```
Comment (hoangdat) [inline lib/jmap/core/method/request/query_method.dart:12]: why the abstract `QueryMethod` need `FilterAsTree` and `SortAsTree`?

File: lib/jmap/mail/mailbox/mailbox.dart [hunk 7]
```diff
@@ -180,3 +180,12 @@ class IsSubscribed with EquatableMixin {
...
 }
+
+class HasAnyRole with EquatableMixin {
+  final bool value;
+
+  HasAnyRole(this.value);
+
+  @override
+  List<Object?> get props => [value];
+}
```
Comment (hoangdat) [inline lib/jmap/mail/mailbox/mailbox.dart:191]: Why JMAP entity need to know about the properties of `Query` request?

File: lib/jmap/mail/mailbox/mailbox_filter_condition.dart [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+class MailboxFilterCondition extends FilterCondition {
+
+  @JsonKey(includeIfNull: false)
+  final Role? role;
+  @JsonKey(includeIfNull: false)
+  final MailboxName? name;
+  @JsonKey(includeIfNull: false)
+  final bool? hasAnyRole;
+  @JsonKey(includeIfNull: false)
+  final IsSubscribed? isSubscribed;
```
Comment (hoangdat) [inline lib/jmap/mail/mailbox/mailbox_filter_condition.dart:27]: still need the strong-type for `IsSubscribed`?

File: lib/jmap/mail/mailbox/query/query_mailbox_method.dart [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  @override
+  Set<CapabilityIdentifier> get requiredCapabilities => {
+    CapabilityIdentifier.jmapCore,
+    CapabilityIdentifier.jmapMail
+  };
+
+  @override
+  List<Object?> get props => [methodName, accountId, filter, limit];
+
+  factory QueryMailboxMethod.fromJson(Map<String, dynamic> json) {
```
Comment (hoangdat) [inline lib/jmap/mail/mailbox/query/query_mailbox_method.dart:26]: What is the reason to handle `fromJson` and `toJson` by your-self?

File: lib/jmap/mail/mailbox/mailbox_filter_condition.dart [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    this.role,
+    this.parentId,
+    this.name,
+    this.hasAnyRole,
+    this.isSubscribed,
+  });
+
+  @override
+  List<Object?> get props => [
+    role,
```
Comment (dab246) [inline lib/jmap/mail/mailbox/mailbox_filter_condition.dart:39]: Should add more property at here

File: lib/jmap/mail/mailbox/query/query_mailbox_response.dart [hunk 11]
```diff
@@ -0,0 +1,43 @@
+
+
```
Comment (dab246) [inline lib/jmap/mail/mailbox/query/query_mailbox_response.dart:2]: remove excess line

File: test/jmap/mailbox/query/query_mailbox_test.dart [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+        ));
+      final getMailboxInvocation = jmapRequestBuilder
+          .invocation(getMailBoxMethod, methodCallId: MethodCallId('c3'));
+      final result = await (jmapRequestBuilder
+            ..usings(getMailBoxMethod.requiredCapabilities))
+          .build()
+          .execute();
+
+      final resultList = result.parse<GetMailboxResponse>(
+          getMailboxInvocation.methodCallId, GetMailboxResponse.deserialize);
```
Comment (dab246) [inline test/jmap/mailbox/query/query_mailbox_test.dart:158]: each param should on 1 line

File: test/jmap/mailbox/query/query_mailbox_test.dart [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+      final processingInvocation = ProcessingInvocation();
+      final jmapRequestBuilder =
+          JmapRequestBuilder(httpClient, processingInvocation);
+      final accountId = AccountId(Id(
+          '0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555'));
+      final queryMailboxMethod = QueryMailboxMethod(accountId)
+        ..addFilters(MailboxFilterCondition(role: Role('Spam')))
+        ..addLimit(UnsignedInt(1));
+      final queryMailboxInvocation = jmapRequestBuilder
+          .invocation(queryMailboxMethod, methodCallId: MethodCallId('c2'));
```
Comment (dab246) [inline test/jmap/mailbox/query/query_mailbox_test.dart:143]: each param should on 1 line

File: test/jmap/mailbox/query/query_mailbox_test.dart [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+          },
+          headers: {
+            "accept": "application/json;jmapVersion=rfc-8621",
+            "content-type": "application/json; charset=utf-8"
+          });
+
+      final httpClient = HttpClient(dio);
+      final processingInvocation = ProcessingInvocation();
+      final jmapRequestBuilder =
+          JmapRequestBuilder(httpClient, processingInvocation);
```
Comment (dab246) [inline test/jmap/mailbox/query/query_mailbox_test.dart:136]: each param should on 1 line

File: lib/jmap/mail/email/email_filter_condition.g.dart [hunk 15]
```diff
@@ -6,36 +6,38 @@ part of 'email_filter_condition.dart';
...
-    from: json['from'] as String?,
-    to: json['to'] as String?,
-    cc: json['cc'] as String?,
-    bcc: json['bcc'] as String?,
-    subject: json['subject'] as String?,
-    body: json['body'] as String?,
-    header: (json['header'] as List<dynamic>?)?.map((e) => e as String).toSet(),
-  );
-}
+EmailFilterCondition _$EmailFilterConditionFromJson(
```
Comment (hoangdat) [inline lib/jmap/mail/email/email_filter_condition.g.dart:9]: why we need this change? any impact to EmailFilterCondition?

**Review by dab246 — APPROVED** [review]

File: lib/jmap/mail/email/email_filter_condition.g.dart (hunk 15 above)
Comment (dab246) [reply lib/jmap/mail/email/email_filter_condition.g.dart:9]: This I think the IDE format itself when we run `flutter build run`

**Review by hoangdat — APPROVED** [review]

**Pull request merged**